pkg/sriov/sriov.go

```diff
@@ -67,108 +68,134 @@ func NewSriovManager() Manager {
 	linkName := conf.OrigVfState.HostIFName
 
-	linkObj, err := s.nLink.LinkByName(linkName)
+	tempNS, err := ns.TempNetNS()
 	if err != nil {
-		return fmt.Errorf("error getting VF netdevice with name %s", linkName)
+		return fmt.Errorf("failed to create tempNS: %v", err)
 	}
+	defer tempNS.Close()
 
+	linkObj, err := s.nLink.LinkByName(linkName)
+	if err != nil {
+		return fmt.Errorf("error: %v. Failed to get VF netdevice with name %s", err, linkName)
+	}
+	if linkObj.Attrs().Flags&net.FlagUp == net.FlagUp {
+		defer func() {
+			if err != nil {
+				logging.Debug("Setting link up in defer because of ", "error", err)
+				if linkObj, err := s.nLink.LinkByName(linkName); err == nil {
+					_ = s.nLink.LinkSetUp(linkObj)
+				}
+			}
+		}()
+	}
 	// Save the original effective MAC address before overriding it
 	conf.OrigVfState.EffectiveMAC = linkObj.Attrs().HardwareAddr.String()
 
-	// tempName used as intermediary name to avoid name conflicts
-	tempName := fmt.Sprintf("%s%d", "temp_", linkObj.Attrs().Index)
-
-	// 1. Set link down
-	logging.Debug("1. Set link down",
+	// 1.Move interface to tempNS
+	logging.Debug("1. Move the interface to tempNS",
 		"func", "SetupVF",
 		"linkObj", linkObj)
-	if err := s.nLink.LinkSetDown(linkObj); err != nil {
-		return fmt.Errorf("failed to down vf device %q: %v", linkName, err)
-	}
-
-	// 2. Set temp name
-	logging.Debug("2. Set temp name",
-		"func", "SetupVF",
-		"linkObj", linkObj,
-		"tempName", tempName)
-	if err := s.nLink.LinkSetName(linkObj, tempName); err != nil {
-		return fmt.Errorf("error setting temp IF name %s for %s", tempName, linkName)
-	}
-
-	// 3. Remove alt name from the nic
-	logging.Debug("3. Remove interface original name from alt names",
-		"func", "SetupVF",
-		"linkObj", linkObj,
-		"OriginalLinkName", linkName,
-		"tempName", tempName)
-	linkObj, err = s.nLink.LinkByName(tempName)
-	if err != nil {
-		return fmt.Errorf("error getting VF netdevice with name %s: %v", tempName, err)
+	if err = s.nLink.LinkSetNsFd(linkObj, int(tempNS.Fd())); err != nil {
+		return fmt.Errorf("failed to move %q to tempNS: %v", linkName, err)
 	}
-	for _, altName := range linkObj.Attrs().AltNames {
-		if altName == linkName {
-			if err := s.nLink.LinkDelAltName(linkObj, linkName); err != nil {
-				return fmt.Errorf("error removing VF altname %s: %v", linkName, err)
+	if err = tempNS.Do(func(linkNS ns.NetNS) error {
+		// lookup the device in tempNS (index might have changed)
+		tempNSLinkObj, err := s.nLink.LinkByName(linkName)
+		if err != nil {
+			return fmt.Errorf("failed to find %q in tempNS: %v", linkName, err)
+		}
+		defer func() {
+			if err != nil {
+				logging.Debug("Move the interface back to initNS because of ", "error", err)
+				initns, _ := ns.GetCurrentNS()
+				linkSetNsFdError := netlink.LinkSetNsFd(linkObj, int(initns.Fd()))
+				logging.Debug("LinkSetNsFd failed in defer", "error", linkSetNsFdError)
```

Comment:
log this conditionally if linkSetNsFdError fail ?

Reply:
done